{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>##description: evidence-based annotation of the human genome (GRCh38), version 26 (Ensembl 88)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>##provider: GENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>##contact: gencode-help@sanger.ac.uk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>##format: gtf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>##date: 2017-03-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>##collapsed version generated by GTEx LDACC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ##description: evidence-based annotation of the human genome (GRCh38), version 26 (Ensembl 88)\n",
       "0                                ##provider: GENCODE                                            \n",
       "1               ##contact: gencode-help@sanger.ac.uk                                            \n",
       "2                                      ##format: gtf                                            \n",
       "3                                 ##date: 2017-03-14                                            \n",
       "4        ##collapsed version generated by GTEx LDACC                                            "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"/storage/ydong/data/gtexRNA/gencode.v26.GRCh38.genes.gtf\",sep = \"\\t\", nrows = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/storage/ydong/data/gtexRNA/gencode.v26.GRCh38.genes.gtf\",sep = \"\\t\", skiprows = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>##collapsed version generated by GTEx LDACC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chr1</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">HAVANA</th>\n",
       "      <th>gene</th>\n",
       "      <th>11869</th>\n",
       "      <th>14403</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <th>11869</th>\n",
       "      <th>14403</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">exon</th>\n",
       "      <th>11869</th>\n",
       "      <th>12227</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12613</th>\n",
       "      <th>12721</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12975</th>\n",
       "      <th>13052</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chrM</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">ENSEMBL</th>\n",
       "      <th>transcript</th>\n",
       "      <th>15888</th>\n",
       "      <th>15953</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <th>15888</th>\n",
       "      <th>15953</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>15956</th>\n",
       "      <th>16023</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <th>15956</th>\n",
       "      <th>16023</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <th>15956</th>\n",
       "      <th>16023</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>441071 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ##collapsed version generated by GTEx LDACC\n",
       "chr1 HAVANA  gene       11869 14403 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "             transcript 11869 14403 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "             exon       11869 12227 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "                        12613 12721 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "                        12975 13052 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "...                                                                                      ...\n",
       "chrM ENSEMBL transcript 15888 15953 . + .  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...\n",
       "             exon       15888 15953 . + .  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...\n",
       "             gene       15956 16023 . - .  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...\n",
       "             transcript 15956 16023 . - .  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...\n",
       "             exon       15956 16023 . - .  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...\n",
       "\n",
       "[441071 rows x 1 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract exon annotation only\n",
    "idx = pd.IndexSlice\n",
    "df = df.loc[idx[:,:,\"exon\"],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>##collapsed version generated by GTEx LDACC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chr1</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">HAVANA</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">exon</th>\n",
       "      <th>11869</th>\n",
       "      <th>12227</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12613</th>\n",
       "      <th>12721</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12975</th>\n",
       "      <th>13052</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13221</th>\n",
       "      <th>14403</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29534</th>\n",
       "      <th>29553</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chrM</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">ENSEMBL</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">exon</th>\n",
       "      <th>14149</th>\n",
       "      <th>14673</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000198695.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14674</th>\n",
       "      <th>14742</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210194.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14747</th>\n",
       "      <th>15887</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000198727.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15888</th>\n",
       "      <th>15953</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210195.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15956</th>\n",
       "      <th>16023</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_id \"ENSG00000210196.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328671 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           ##collapsed version generated by GTEx LDACC\n",
       "chr1 HAVANA  exon 11869 12227 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "                  12613 12721 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "                  12975 13052 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "                  13221 14403 . + .  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...\n",
       "                  29534 29553 . - .  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...\n",
       "...                                                                                ...\n",
       "chrM ENSEMBL exon 14149 14673 . - .  gene_id \"ENSG00000198695.2\"; transcript_id \"EN...\n",
       "                  14674 14742 . - .  gene_id \"ENSG00000210194.1\"; transcript_id \"EN...\n",
       "                  14747 15887 . + .  gene_id \"ENSG00000198727.2\"; transcript_id \"EN...\n",
       "                  15888 15953 . + .  gene_id \"ENSG00000210195.2\"; transcript_id \"EN...\n",
       "                  15956 16023 . - .  gene_id \"ENSG00000210196.2\"; transcript_id \"EN...\n",
       "\n",
       "[328671 rows x 1 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# chromosome number\n",
    "chromNum = list(df.index.get_level_values(0))\n",
    "# transform from \"chrX\" to \"X\"\n",
    "for i in range(len(chromNum)):\n",
    "    chromNum[i] = chromNum[i].split(\"chr\")[-1]\n",
    "    \n",
    "# start position\n",
    "start = list(df.index.get_level_values(3))\n",
    "# end position\n",
    "end = list(df.index.get_level_values(4))\n",
    "# forward/reverse strand\n",
    "strand = list(df.index.get_level_values(6))\n",
    "\n",
    "#extract exonID from the values\n",
    "exonID = list(df.values)\n",
    "for i in range(len(exonID)):\n",
    "    curr = exonID[i][0]\n",
    "    curr = curr.strip().split(\";\")[-3].split(\"\\\"\")[1]\n",
    "    exonID[i] = curr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>probe.id</th>\n",
       "      <th>probe.chr</th>\n",
       "      <th>probe.start</th>\n",
       "      <th>probe.stop</th>\n",
       "      <th>probe.strand</th>\n",
       "      <th>gene.id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5_1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5_2</td>\n",
       "      <td>1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000223972.5_3</td>\n",
       "      <td>1</td>\n",
       "      <td>12975</td>\n",
       "      <td>13052</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000223972.5_4</td>\n",
       "      <td>1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14403</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000227232.5_1</td>\n",
       "      <td>1</td>\n",
       "      <td>29534</td>\n",
       "      <td>29553</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328666</th>\n",
       "      <td>ENSG00000198695.2_1</td>\n",
       "      <td>M</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328667</th>\n",
       "      <td>ENSG00000210194.1_1</td>\n",
       "      <td>M</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328668</th>\n",
       "      <td>ENSG00000198727.2_1</td>\n",
       "      <td>M</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328669</th>\n",
       "      <td>ENSG00000210195.2_1</td>\n",
       "      <td>M</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328670</th>\n",
       "      <td>ENSG00000210196.2_1</td>\n",
       "      <td>M</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328671 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   probe.id probe.chr  probe.start  probe.stop probe.strand  \\\n",
       "0       ENSG00000223972.5_1         1        11869       12227            +   \n",
       "1       ENSG00000223972.5_2         1        12613       12721            +   \n",
       "2       ENSG00000223972.5_3         1        12975       13052            +   \n",
       "3       ENSG00000223972.5_4         1        13221       14403            +   \n",
       "4       ENSG00000227232.5_1         1        29534       29553            -   \n",
       "...                     ...       ...          ...         ...          ...   \n",
       "328666  ENSG00000198695.2_1         M        14149       14673            -   \n",
       "328667  ENSG00000210194.1_1         M        14674       14742            -   \n",
       "328668  ENSG00000198727.2_1         M        14747       15887            +   \n",
       "328669  ENSG00000210195.2_1         M        15888       15953            +   \n",
       "328670  ENSG00000210196.2_1         M        15956       16023            -   \n",
       "\n",
       "                  gene.id  \n",
       "0       ENSG00000223972.5  \n",
       "1       ENSG00000223972.5  \n",
       "2       ENSG00000223972.5  \n",
       "3       ENSG00000223972.5  \n",
       "4       ENSG00000227232.5  \n",
       "...                   ...  \n",
       "328666  ENSG00000198695.2  \n",
       "328667  ENSG00000210194.1  \n",
       "328668  ENSG00000198727.2  \n",
       "328669  ENSG00000210195.2  \n",
       "328670  ENSG00000210196.2  \n",
       "\n",
       "[328671 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame (list(zip(exonID,chromNum,start,end,strand)),columns= [\"probe.id\",\"probe.chr\",\"probe.start\",\"probe.stop\",\"probe.strand\"])\n",
    "df[\"gene.id\"] = df[\"probe.id\"].apply(lambda x:x.split(\"_\")[0])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"/storage/ydong/data/gtexRNA/gencode_gene_annotations_GRCh38.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
