{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We only perform our analysis on chr1 ~ chr 22, as sex chromosome expression varies between male and female, would lead to complex pipelines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = \"/gymreklab-tscc/ydong/data/allTis/\"\n",
    "tisSuffix = os.listdir(OUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cervix-Endocervix\n",
      "Skin-SunExposed(Lowerleg)\n",
      "Thyroid\n",
      "Artery-Tibial\n",
      "Esophagus-Muscularis\n",
      "Testis\n",
      "Vagina\n",
      "Esophagus-Mucosa\n",
      "Skin-NotSunExposed(Suprapubic)\n",
      "Artery-Aorta\n",
      "Stomach\n",
      "Brain-Caudate(basalganglia)\n",
      "Cells-Leukemiacellline(CML)\n",
      "Breast-MammaryTissue\n",
      "Heart-AtrialAppendage\n",
      "Brain-Cerebellum\n",
      "Brain-Anteriorcingulatecortex(BA24)\n",
      "Pancreas\n",
      "Adipose-Subcutaneous\n",
      "MinorSalivaryGland\n",
      "Liver\n",
      "Bladder\n",
      "Esophagus-GastroesophagealJunction\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "for currTis in tisSuffix:\n",
    "    try:\n",
    "        os.rename(OUT_DIR+currTis+'/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads.par_Parsed.csv',\n",
    "                  OUT_DIR+currTis+'/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads_Parsed.csv')\n",
    "    except OSError:\n",
    "        print(currTis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aaa\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "try:\n",
    "    os.rename(\"fasdfadsfasd.txt\",\"dfasdfas.txt\")\n",
    "except OSError:\n",
    "    print(\"aaa\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXONFILE = OUT_DIR + tisSuffix[1]+'/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads_Parsed.csv'#args.exon_expr\n",
    "JUNCTIONFILE =OUT_DIR + tisSuffix[1]+ '/GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions_Parsed.csv'#args.junction_expr\n",
    "ANNOTFILE = '/storage/ydong/data/gtexRNA/gencode_gene_annotations_GRCh38.csv'#args.exprannot\n",
    "    \n",
    "    \n",
    "annot_df_t = pd.read_csv(ANNOTFILE,index_col = 0,dtype = {\"probe.chr\":str})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_df_t = pd.read_csv(EXONFILE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Description</th>\n",
       "      <th>GTEX-111CU-0426-SM-5GZY1</th>\n",
       "      <th>GTEX-111FC-1326-SM-5N9D9</th>\n",
       "      <th>GTEX-1122O-0526-SM-5N9DM</th>\n",
       "      <th>GTEX-117YX-2126-SM-5GIEL</th>\n",
       "      <th>GTEX-11DXX-0726-SM-5H12X</th>\n",
       "      <th>GTEX-11EQ9-0426-SM-5A5JY</th>\n",
       "      <th>GTEX-11LCK-0126-SM-5A5M5</th>\n",
       "      <th>GTEX-11NSD-0426-SM-5N9CR</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZP4G-0326-SM-4YCEF</th>\n",
       "      <th>GTEX-ZPCL-0626-SM-DNZZ6</th>\n",
       "      <th>GTEX-ZPIC-0126-SM-DNZZ9</th>\n",
       "      <th>GTEX-ZPU1-1026-SM-4YCEQ</th>\n",
       "      <th>GTEX-ZTTD-1726-SM-57WEL</th>\n",
       "      <th>GTEX-ZV6S-0826-SM-5NQ6Z</th>\n",
       "      <th>GTEX-ZVP2-0526-SM-51MSC</th>\n",
       "      <th>GTEX-ZVZP-0726-SM-59HKA</th>\n",
       "      <th>GTEX-ZYFG-0726-SM-5GIDX</th>\n",
       "      <th>GTEX-ZZPU-0126-SM-5E446</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.118421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.881579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657337</th>\n",
       "      <td>ENSG00000284600.1_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657338</th>\n",
       "      <td>ENSG00000284600.1_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657339</th>\n",
       "      <td>ENSG00000284600.1_5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657340</th>\n",
       "      <td>ENSG00000284600.1_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657341</th>\n",
       "      <td>ENSG00000284600.1_7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>657342 rows × 243 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Unnamed: 0 Description  GTEX-111CU-0426-SM-5GZY1  \\\n",
       "0                         0     DDX11L1                       0.0   \n",
       "1                         1     DDX11L1                       0.0   \n",
       "2                         2     DDX11L1                       0.0   \n",
       "3                         3     DDX11L1                       1.0   \n",
       "4                         4      WASH7P                       0.0   \n",
       "...                     ...         ...                       ...   \n",
       "657337  ENSG00000284600.1_3         NaN                       NaN   \n",
       "657338  ENSG00000284600.1_4         NaN                       NaN   \n",
       "657339  ENSG00000284600.1_5         NaN                       NaN   \n",
       "657340  ENSG00000284600.1_6         NaN                       NaN   \n",
       "657341  ENSG00000284600.1_7         NaN                       NaN   \n",
       "\n",
       "        GTEX-111FC-1326-SM-5N9D9  GTEX-1122O-0526-SM-5N9DM  \\\n",
       "0                            0.0                       0.0   \n",
       "1                            0.0                       0.0   \n",
       "2                            0.0                       0.0   \n",
       "3                            1.0                       0.0   \n",
       "4                            0.0                       0.0   \n",
       "...                          ...                       ...   \n",
       "657337                       NaN                       NaN   \n",
       "657338                       NaN                       NaN   \n",
       "657339                       NaN                       NaN   \n",
       "657340                       NaN                       NaN   \n",
       "657341                       NaN                       NaN   \n",
       "\n",
       "        GTEX-117YX-2126-SM-5GIEL  GTEX-11DXX-0726-SM-5H12X  \\\n",
       "0                            0.0                       0.0   \n",
       "1                            0.0                       0.0   \n",
       "2                            0.0                       0.0   \n",
       "3                            0.0                       0.0   \n",
       "4                            0.0                       0.0   \n",
       "...                          ...                       ...   \n",
       "657337                       NaN                       NaN   \n",
       "657338                       NaN                       NaN   \n",
       "657339                       NaN                       NaN   \n",
       "657340                       NaN                       NaN   \n",
       "657341                       NaN                       NaN   \n",
       "\n",
       "        GTEX-11EQ9-0426-SM-5A5JY  GTEX-11LCK-0126-SM-5A5M5  \\\n",
       "0                            0.0                       0.0   \n",
       "1                            0.0                       0.0   \n",
       "2                            0.0                       0.0   \n",
       "3                            1.0                       0.0   \n",
       "4                            0.0                       0.0   \n",
       "...                          ...                       ...   \n",
       "657337                       NaN                       NaN   \n",
       "657338                       NaN                       NaN   \n",
       "657339                       NaN                       NaN   \n",
       "657340                       NaN                       NaN   \n",
       "657341                       NaN                       NaN   \n",
       "\n",
       "        GTEX-11NSD-0426-SM-5N9CR  ...  GTEX-ZP4G-0326-SM-4YCEF  \\\n",
       "0                            0.0  ...                      1.0   \n",
       "1                            0.0  ...                      0.0   \n",
       "2                            0.0  ...                      0.0   \n",
       "3                            0.0  ...                      0.0   \n",
       "4                            0.0  ...                      0.0   \n",
       "...                          ...  ...                      ...   \n",
       "657337                       NaN  ...                      NaN   \n",
       "657338                       NaN  ...                      NaN   \n",
       "657339                       NaN  ...                      NaN   \n",
       "657340                       NaN  ...                      NaN   \n",
       "657341                       NaN  ...                      NaN   \n",
       "\n",
       "        GTEX-ZPCL-0626-SM-DNZZ6  GTEX-ZPIC-0126-SM-DNZZ9  \\\n",
       "0                           0.0                      0.0   \n",
       "1                           0.0                      0.0   \n",
       "2                           0.0                      0.0   \n",
       "3                           0.0                      0.0   \n",
       "4                           0.0                      0.0   \n",
       "...                         ...                      ...   \n",
       "657337                      NaN                      NaN   \n",
       "657338                      NaN                      NaN   \n",
       "657339                      NaN                      NaN   \n",
       "657340                      NaN                      NaN   \n",
       "657341                      NaN                      NaN   \n",
       "\n",
       "        GTEX-ZPU1-1026-SM-4YCEQ  GTEX-ZTTD-1726-SM-57WEL  \\\n",
       "0                           0.0                      1.0   \n",
       "1                           0.0                      0.0   \n",
       "2                           0.0                      0.0   \n",
       "3                           2.0                      0.0   \n",
       "4                           0.0                      0.0   \n",
       "...                         ...                      ...   \n",
       "657337                      NaN                      NaN   \n",
       "657338                      NaN                      NaN   \n",
       "657339                      NaN                      NaN   \n",
       "657340                      NaN                      NaN   \n",
       "657341                      NaN                      NaN   \n",
       "\n",
       "        GTEX-ZV6S-0826-SM-5NQ6Z  GTEX-ZVP2-0526-SM-51MSC  \\\n",
       "0                           0.0                      0.0   \n",
       "1                           0.0                      0.0   \n",
       "2                           0.0                      0.0   \n",
       "3                           4.0                      2.0   \n",
       "4                           0.0                      0.0   \n",
       "...                         ...                      ...   \n",
       "657337                      NaN                      NaN   \n",
       "657338                      NaN                      NaN   \n",
       "657339                      NaN                      NaN   \n",
       "657340                      NaN                      NaN   \n",
       "657341                      NaN                      NaN   \n",
       "\n",
       "        GTEX-ZVZP-0726-SM-59HKA  GTEX-ZYFG-0726-SM-5GIDX  \\\n",
       "0                           0.0                      0.0   \n",
       "1                           0.0                      0.0   \n",
       "2                           0.0                      0.0   \n",
       "3                           0.0                      0.0   \n",
       "4                           0.0                      0.0   \n",
       "...                         ...                      ...   \n",
       "657337                      NaN                      NaN   \n",
       "657338                      NaN                      NaN   \n",
       "657339                      NaN                      NaN   \n",
       "657340                      NaN                      NaN   \n",
       "657341                      NaN                      NaN   \n",
       "\n",
       "        GTEX-ZZPU-0126-SM-5E446  \n",
       "0                      2.000000  \n",
       "1                      0.118421  \n",
       "2                      0.000000  \n",
       "3                      4.881579  \n",
       "4                      0.000000  \n",
       "...                         ...  \n",
       "657337                      NaN  \n",
       "657338                      NaN  \n",
       "657339                      NaN  \n",
       "657340                      NaN  \n",
       "657341                      NaN  \n",
       "\n",
       "[657342 rows x 243 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_df_t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in files\n",
    "exon_df_t = pd.read_csv(EXONFILE,index_col = \"Name\")\n",
    "\n",
    "junction_df_t = pd.read_csv(JUNCTIONFILE,index_col = \"Name\",dtype = {\"Chr\":str})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['Unnamed: 0'] not found in axis\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-4225ad16aecb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    143\u001b[0m         \u001b[0mjunction_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjunction_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mjunction_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Chr\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCHROM\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0mexon_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexon_df_t\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannot_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'probe.id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 145\u001b[0;31m         \u001b[0mexon_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexon_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Unnamed: 0\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maxis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    146\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    147\u001b[0m         \u001b[0;31m#removing duplicated samples by taking highest coverage (in exon expression) sample\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4115\u001b[0m             \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4116\u001b[0m             \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minplace\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4117\u001b[0;31m             \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4118\u001b[0m         )\n\u001b[1;32m   4119\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   3912\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32min\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3913\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3914\u001b[0;31m                 \u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_drop_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3915\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3916\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_drop_axis\u001b[0;34m(self, labels, axis, level, errors)\u001b[0m\n\u001b[1;32m   3944\u001b[0m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3945\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3946\u001b[0;31m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3947\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0maxis_name\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mnew_axis\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3948\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, errors)\u001b[0m\n\u001b[1;32m   5338\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5339\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0merrors\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5340\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"{} not found in axis\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5341\u001b[0m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5342\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['Unnamed: 0'] not found in axis\""
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Compute Percent Spliced In (PSI)\n",
    "# PSI_df: the output matrix of psi\n",
    "# exon_gene_df: exon read count matrix\n",
    "# junction_gene_df: Exon-exon junction read count matrix\n",
    "# annot_gene_df: annotation file for exons\n",
    "def compute_gene_PSI(PSI_df,exon_gene_df,junction_gene_df,annot_gene_df,read_len):\n",
    "     \n",
    "    exons = list(exon_gene_df.index)\n",
    "    min_reads = 10\n",
    "    \n",
    "    # loop through each exon\n",
    "    for exon in exons:\n",
    "\n",
    "        exon_chrom = annot_gene_df.loc[exon]['probe.chr']\n",
    "        exon_start = annot_gene_df.loc[exon]['probe.start']\n",
    "        exon_stop = annot_gene_df.loc[exon]['probe.stop']\n",
    "        exon_len = math.fabs(exon_start - exon_stop) + 1\n",
    "        exon_strand = annot_gene_df.loc[exon]['probe.strand']\n",
    "        if exon_strand == '-':\n",
    "            orig_start = exon_start\n",
    "            exon_start = exon_stop\n",
    "            exon_stop = orig_start\n",
    "        \n",
    "        \n",
    "        #Get the A,B,C reads for PSI calculation\n",
    "        B_intervals = []\n",
    "        C_intervals = []\n",
    "        for interval in junction_gene_df.index:\n",
    "            chrom,start,stop = interval.split('_')\n",
    "            chrom = chrom.split(\"chr\")[1]\n",
    "            if str(chrom) != str(exon_chrom):\n",
    "                print(chrom)\n",
    "                print(exon_chrom)\n",
    "                print('Chromosomes aren\\'t matching, something is wrong!!')\n",
    "\n",
    "            if int(start) < exon_start and int(stop) > exon_stop:#C\n",
    "                C_intervals.append(interval)\n",
    "            if int(start) < exon_start and (int(stop) >= exon_start and int(stop) <= exon_stop):#B1\n",
    "                B_intervals.append(interval)\n",
    "            if (int(start) >= exon_start and int(start) <= exon_stop) and int(stop) > exon_stop:#B2\n",
    "                B_intervals.append(interval)\n",
    "\n",
    "        A_reads = exon_gene_df.loc[exon]\n",
    "        B_reads = junction_gene_df.loc[B_intervals]\n",
    "        B_reads = B_reads.iloc[:,2:].sum()\n",
    "\n",
    "        C_reads = junction_gene_df.loc[C_intervals]\n",
    "        C_reads = C_reads.iloc[:,2:].sum()\n",
    "\n",
    "        total_reads = A_reads + B_reads + C_reads\n",
    "        if len(total_reads[total_reads < min_reads])/len(total_reads) > 0.2:\n",
    "            continue\n",
    "        else:\n",
    "            A_B_norm = (A_reads + B_reads)/(read_len + exon_len-1)\n",
    "            C_norm = C_reads/(read_len-1)\n",
    "\n",
    "            mask = (A_B_norm == 0) & (C_norm == 0)\n",
    "            PSI_norm = 100*((A_B_norm)/(A_B_norm + C_norm)) if (A_B_norm.sum() + C_norm.sum()) > 0.00 else 0\n",
    "            PSI_norm = PSI_norm.where(~mask,0)\n",
    "            PSI_df[exon] = PSI_norm\n",
    "\n",
    "    return PSI_df\n",
    "\n",
    "\n",
    "# find out the maximum coverage sample for each individual\n",
    "def max_coverage_samples(exon_df):\n",
    "    'Given exon expression dataframe'\n",
    "    'Take first two elements of sample name'\n",
    "    'If there is duplicate sample'\n",
    "    'Return list of max coverage non-dup samples'\n",
    "    \n",
    "    med_exon_df = pd.DataFrame(exon_df.median(axis=0))\n",
    "    med_exon_df['samples'] = [\"-\".join(i.split('-')[:2]) for i in med_exon_df.index]\n",
    "    med_exon_df['original_samples'] = med_exon_df.index\n",
    "    highest_cov_samples = list(med_exon_df.groupby('samples').idxmax()[0]) \n",
    "    \n",
    "    return highest_cov_samples\n",
    "\n",
    "def filter_exons(PSI_df):\n",
    "    'Given PSI dataframe'\n",
    "    'Filter out exons with little variation:'\n",
    "    'PSI = 100% or PSI = 0% in >90% patients'\n",
    "    'Return filtered PSI'\n",
    "    \n",
    "    PSI_filtered_df = PSI_df.dropna(axis=1)\n",
    "    exons = PSI_filtered_df.columns\n",
    "    exons_to_remove = []\n",
    "    num_samples = len(PSI_df.index)\n",
    "    for exon in exons:\n",
    "        exon_PSI = PSI_filtered_df[exon]\n",
    "        num_zero = len(exon_PSI[exon_PSI == 0])\n",
    "        num_hundred = len(exon_PSI[exon_PSI == 100])\n",
    "        if num_zero/num_samples > 0.9 or num_hundred/num_samples > 0.9:\n",
    "            exons_to_remove.append(exon)\n",
    "    PSI_filtered_df = PSI_filtered_df.drop(exons_to_remove,axis=1).dropna()\n",
    "    new_index = [\"-\".join(i.split('-')[:2]) for i in PSI_filtered_df.index]\n",
    "    PSI_filtered_df.index = new_index\n",
    "\n",
    "    return PSI_filtered_df\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "#     parser = argparse.ArgumentParser(description=\"Compute Percent Spliced In (PSI) for any exon\")\n",
    "#     parser.add_argument(\"--exon_expr\", help=\"File path to exon expression dataset\", type=str, required=True)\n",
    "#     parser.add_argument(\"--junction_expr\", help=\"File path to junction expression dataset\", type=str, required=True)\n",
    "#     parser.add_argument(\"--exprannot\", help=\"Exon expression annotation file\", type=str, required=True)\n",
    "#     parser.add_argument(\"--chrom\", help=\"Restrict analysis to this chromosome\", type=str, required=False)\n",
    "#     parser.add_argument(\"--min_samples\", help=\"Require data for this many samples\", type=int, default=0)\n",
    "#     parser.add_argument(\"--read_len\", help=\"Paired end length, default is GTEx read length: 75bp \", type=int, default=75)\n",
    "#     parser.add_argument(\"--out\", help=\"Write data files to this file\", type=str, required=True)\n",
    "\n",
    "\n",
    "#     args = parser.parse_args()\n",
    "    EXONFILE = '/storage/ydong/data/gtexRNA/lung/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads_Parsed.csv'#args.exon_expr\n",
    "    JUNCTIONFILE = '/storage/ydong/data/gtexRNA/lung/GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions_Parsed.csv'#args.junction_expr\n",
    "    ANNOTFILE = '/storage/ydong/data/gtexRNA/gencode_gene_annotations_GRCh38.csv'#args.exprannot\n",
    "    \n",
    "    \n",
    "    # read in files\n",
    "    exon_df_t = pd.read_csv(EXONFILE,index_col = \"Name\")\n",
    "    annot_df_t = pd.read_csv(ANNOTFILE,index_col = 0,dtype = {\"probe.chr\":str})\n",
    "    junction_df_t = pd.read_csv(JUNCTIONFILE,index_col = \"Name\",dtype = {\"Chr\":str})\n",
    "    \n",
    "    chrList = list(range(2,23))\n",
    "    \n",
    "    for CHROM in chrList:\n",
    "#         CHROM = 1 #args.chrom\n",
    "        MINSAMPLES = 0 #args.min_samples\n",
    "        READLEN = 2*75 #args.read_len\n",
    "        OUTFILE = '/storage/ydong/data/chrom/lung/chr' + str(CHROM) + '_PSI.csv'#args.out\n",
    "\n",
    "        # format CHROM\n",
    "        CHROM = str(CHROM)\n",
    "        annot_df = annot_df_t[annot_df_t['probe.chr'] ==  str(CHROM)]\n",
    "        junction_df =junction_df_t[junction_df_t['Description'].isin(annot_df['gene.id'])]\n",
    "        junction_df = junction_df[junction_df[\"Chr\"] == str(CHROM)]\n",
    "        exon_df = exon_df_t.loc[list(annot_df['probe.id'])]\n",
    "\n",
    "        #removing duplicated samples by taking highest coverage (in exon expression) sample\n",
    "\n",
    "        highest_cov_samples = max_coverage_samples(exon_df) \n",
    "        exon_df = exon_df[highest_cov_samples]\n",
    "        highest_cov_samples[:0] = ['Description','Chr']\n",
    "        junction_df = junction_df[highest_cov_samples]\n",
    "        PSI_df = pd.DataFrame(index=exon_df.columns, columns=exon_df.index)\n",
    "        genes = set(annot_df['gene.id'])\n",
    "        print('Chromosome: {}'.format(CHROM))\n",
    "        print('Exons total: {}'.format(len(exon_df.index)))\n",
    "\n",
    "        for gene in genes:\n",
    "            exon_gene_df = exon_df.loc[exon_df.index.str.startswith(gene)]\n",
    "            junction_gene_df = junction_df[junction_df['Description'] == gene]\n",
    "            annot_gene_df = annot_df[annot_df['gene.id'] == gene]\n",
    "            annot_gene_df.index = annot_gene_df['probe.id']\n",
    "            PSI_df = compute_gene_PSI(PSI_df,\n",
    "                                      exon_gene_df,\n",
    "                                      junction_gene_df,\n",
    "                                      annot_gene_df,\n",
    "                                      READLEN)\n",
    "        PSI_filtered_df = filter_exons(PSI_df.dropna(axis=1))\n",
    "        print('Exons with PSI: {}\\n'.format(len(PSI_filtered_df.columns)))\n",
    "        PSI_filtered_df.to_csv(OUTFILE)\n",
    "        print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(range(2,23))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exons with PSI: 3686\n",
      "\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "for gene in genes:\n",
    "    exon_gene_df = exon_df.loc[exon_df.index.str.startswith(gene)]\n",
    "    junction_gene_df = junction_df[junction_df['Description'] == gene]\n",
    "    annot_gene_df = annot_df[annot_df['gene.id'] == gene]\n",
    "    annot_gene_df.index = annot_gene_df['probe.id']\n",
    "    PSI_df = compute_gene_PSI(PSI_df,\n",
    "                              exon_gene_df,\n",
    "                              junction_gene_df,\n",
    "                              annot_gene_df,\n",
    "                              READLEN)\n",
    "PSI_filtered_df = filter_exons(PSI_df.dropna(axis=1))\n",
    "print('Exons with PSI: {}\\n'.format(len(PSI_filtered_df.columns)))\n",
    "PSI_filtered_df.to_csv(OUTFILE)\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
