{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# python3 Linear.py currTissue CHR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import time \n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "currTissue = sys.argv[1]\n",
    "CHR = sys.argv[2]\n",
    "MINSAMPLE = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Z normalization\n",
    "def ZNorm(vals):\n",
    "    m = np.mean(vals)\n",
    "    sd = np.sqrt(np.var(vals))\n",
    "    if type(sd) == pd.Series:\n",
    "        zVar = True\n",
    "        for x in sd:\n",
    "            if x != 0: zVar = False\n",
    "        if zVar == False:\n",
    "            return (vals-m)/sd, False\n",
    "        else:\n",
    "            return None, True\n",
    "    else:\n",
    "        if sd == 0: return None, True\n",
    "        return [(item-m)/sd for item in vals], False\n",
    "# Linear Regression\n",
    "def LinearRegression(X, Y):\n",
    "    \"\"\"\n",
    "    Perform linear regression, return beta, beta_se, p\n",
    "    \"\"\"\n",
    "    X,returnX = ZNorm(X)\n",
    "    Y,returnY = ZNorm(Y)\n",
    "    if returnX == True or returnY == True or len(X) <= 3:\n",
    "        return None,None,None,None\n",
    "    \n",
    "    X = X.dropna(axis = 1)\n",
    "\n",
    "    X = sm.add_constant(X)\n",
    "    mod_ols = sm.OLS(Y, X, missing='drop')\n",
    "    res_ols = mod_ols.fit()\n",
    "    pval = res_ols.pvalues.loc[\"x1+x2\"]\n",
    "    slope = res_ols.params.loc[\"x1+x2\"]\n",
    "    err = res_ols.bse.loc[\"x1+x2\"]\n",
    "    return res_ols, slope, err, pval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the population PCs\n",
    "df = pd.read_csv(\"/projects/ps-gymreklab/ydong/data/genotypePCA/GTEx_1KG_merged_650.pca.evec\", header = None,skiprows = 1)\n",
    "# df = pd.read_csv(\"/gymreklab-tscc/ydong/data/genotypePCA/GTEx_1KG_merged_650.pca.evec\", header = None,skiprows = 1)\n",
    "# trim white spaces from data\n",
    "df[1] = df[0].apply(lambda x : list(filter(lambda a: a !=\"\" ,x.split(\" \"))))\n",
    "df[2] = df[1].apply(lambda x: x[0])\n",
    "\n",
    "samples = list(df[2])\n",
    "pcDF = pd.DataFrame({\"sample\":samples})\n",
    "\n",
    "for i in range(5):\n",
    "    name = \"PC\" + str(i+1)\n",
    "    pcDF[name] = df[1].apply(lambda x: x[i+1])\n",
    "\n",
    "pcDF = pcDF.set_index(\"sample\")\n",
    "pcDF = pcDF.astype(\"float\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# The chromosome we're interested in studying\n",
    "start = time.time()\n",
    "\n",
    "\n",
    "CHR = str(CHR)\n",
    "print(\"CHR: \" + CHR)\n",
    "DIST = 10000\n",
    "\n",
    "#\n",
    "PSI_Path = \"/projects/ps-gymreklab/ydong/data/allTis/\" +currTissue+\"/PSI/\" + CHR + \"_PSI.csv\"\n",
    "STR_Path = \"/projects/ps-gymreklab/ydong/data/STRs/\" + CHR + \".csv\"\n",
    "annot_Path = \"/projects/ps-gymreklab/ydong/data/gtexRNA/gencode_gene_annotations_GRCh38.csv\"\n",
    "\n",
    "# PSI_Path = \"/gymreklab-tscc/ydong/data/allTis/\" +currTissue+\"/PSI/\" + CHR + \"_PSI.csv\"\n",
    "# STR_Path = \"/gymreklab-tscc/ydong/data/STRs/\" + CHR + \".csv\"\n",
    "# annot_Path = \"/gymreklab-tscc/ydong/data/gtexRNA/gencode_gene_annotations_GRCh38.csv\"\n",
    "\n",
    "\n",
    "\n",
    "PSI_df = pd.read_csv(PSI_Path, index_col = 0)\n",
    "STR_df = pd.read_table(STR_Path,dtype = {\"start\":int})\n",
    "annot_df = pd.read_csv(annot_Path,index_col = \"probe.id\", dtype = {\"probe.chr\":str})\n",
    "annot_df = annot_df[annot_df[\"probe.chr\"] == CHR]\n",
    "\n",
    "peer_df = pd.read_csv(\"/projects/ps-gymreklab/ydong/data/allTis/\"+currTissue + \"/\" + \"peerFactor.csv\",index_col = 0)\n",
    "# peer_df = pd.read_csv(\"/gymreklab-tscc/ydong/data/allTis/\"+currTissue + \"/\" + currTissue+\"PEER.csv\",index_col = 0)\n",
    "\n",
    "\n",
    "peer_df = peer_df.drop(\"V1\",axis = 1)\n",
    "\n",
    "# Find the overlapping GTEx individual for STR data and PSI data\n",
    "strSamples = list(STR_df.columns[2:])\n",
    "psiSamples = list(PSI_df.index)\n",
    "peer_df.index = psiSamples\n",
    "overlapSamples = list(set(strSamples)&set(psiSamples))\n",
    "PSI_df = PSI_df.loc[overlapSamples,:]\n",
    "peer_df = peer_df.loc[overlapSamples,:]\n",
    "STR_df = STR_df [[\"chrom\",\"start\"] +overlapSamples]\n",
    "\n",
    "\n",
    "\n",
    "psiSamples = list(PSI_df.columns)\n",
    "res_df = pd.DataFrame(columns =[ \"chrom\", \"gene\",\"str.id\", \"str.start\", \"n.miss\", \"slope\", \"slope_p\", \"error\",\"error_p\",  \"pVal\",\"pVal_p\",\"start.dist\", \"stop.dist\"] )\n",
    "\n",
    "\n",
    "for gene in psiSamples:\n",
    "    \n",
    "    print(gene)\n",
    "    \n",
    "    geneStart = annot_df.loc[gene,\"probe.start\"]\n",
    "    geneStop = annot_df.loc[gene,\"probe.stop\"]\n",
    "    cis_df  = STR_df[(STR_df[\"start\"] >= geneStart - DIST) & (STR_df[\"start\"] <= geneStop + DIST)]\n",
    "\n",
    "\n",
    "    if len(cis_df) == 0:\n",
    "        continue\n",
    "\n",
    "    # deal with multiple matching of strs and target gene\n",
    "    indList = list(cis_df.index)\n",
    "\n",
    "    for ind in indList:\n",
    "        locus_str = pd.DataFrame(cis_df.loc[ind,cis_df.columns[2:]])\n",
    "        test_str = \"STR_\" + str(cis_df.loc[ind,\"start\"])\n",
    "        locus_str.columns = [test_str]\n",
    "        locus_str['x1'] = locus_str[test_str].apply(lambda x: x.split(',')[0] )\n",
    "        locus_str['x2'] = locus_str[test_str].apply(lambda x: x.split(',')[1] )\n",
    "\n",
    "        # filter out the nan STRs\n",
    "        samples_to_keep = [overlapSamples[k] for k in range(len(overlapSamples)) if str(locus_str.iloc[:,0].values[k]) != \"None\" and  str(locus_str.iloc[:,0].values[k]) != 'NA,NA']\n",
    "        locus_str = locus_str.loc[samples_to_keep,:]\n",
    "        locus_str['x1+x2'] = locus_str[test_str].apply(lambda x: int(x.split(',')[0]) + int(x.split(',')[1]))\n",
    "\n",
    "        # filter out all 0 STRs\n",
    "        if locus_str[\"x1+x2\"].sum() == 0:\n",
    "            continue\n",
    "\n",
    "        exonStartDist = cis_df.loc[ind,\"start\"] - geneStart\n",
    "        exonStopDist = cis_df.loc[ind,\"start\"] - geneStop\n",
    "\n",
    "        # get the PSI matrix for current str and current gene\n",
    "        currPSI = PSI_df.loc[samples_to_keep,gene]\n",
    "\n",
    "        # get the covariate Population PCs\n",
    "        currpcDF = pcDF.loc[samples_to_keep]\n",
    "\n",
    "        # get the covariate PEER factors\n",
    "        currPeer = peer_df.loc[samples_to_keep]\n",
    "\n",
    "        # add in covariates for the str sums\n",
    "        xDF = pd.concat([locus_str[\"x1+x2\"],currpcDF,currPeer],axis = 1)\n",
    "\n",
    "\n",
    "        # linear regression\n",
    "        res_ols, slope, err, pval = LinearRegression(xDF, currPSI)\n",
    "\n",
    "        if res_ols == None and slope == None and err == None and pval == None:\n",
    "            continue\n",
    "        \n",
    "         # Permutation Test to validate statistical significance\n",
    "        currPSI_p = random.sample(list(currPSI),len(currPSI))\n",
    "        res_ols_p, slope_p, err_p, pval_p = LinearRegression(xDF, currPSI_p)\n",
    "        \n",
    "        \n",
    "        res_df = res_df.append({\"chrom\":CHR,\"gene\":gene,\"str.id\":test_str,\"str.start\":cis_df.loc[ind,\"start\"],\n",
    "                                \"n.miss\": len(STR_df.columns) -2-len(locus_str), \"slope\":slope,\"slope_p\":slope_p,\n",
    "                                \"error\":err,\"error_p\":err_p,\"pVal\":pval,\"pVal_p\":pval_p, \"start.dist\":exonStartDist, \n",
    "                                \"stop.dist\":exonStopDist},ignore_index=True)\n",
    "\n",
    "res_df.to_csv(\"/projects/ps-gymreklab/ydong/data/allTis/\" +currTissue+\"/result/chr\" + str(CHR) + \"Res.csv\")\n",
    "\n",
    "end = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(end - start)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
