{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "import time\n",
    "import pandas as pd\n",
    "import math\n",
    "import numpy as np\n",
    "import sys\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the vcf file\n",
    "vcf_reader = vcf.Reader(open(\"/storage/ydong/data/Filter_Merged_STRs_All_Samples_New.vcf.gz\", mode = \"rb\"))\n",
    "newPos = pd.read_csv(\"/storage/ydong/data/hg38.bed\",sep = \"\\t\",header = None)\n",
    "newPos.columns = [\"CHR\",\"start\",\"end\"]\n",
    "oldPos = pd.read_csv(\"/storage/ydong/data/hg19.bed\",sep = \" \",header = None)\n",
    "oldPos.columns = [\"CHR\",\"start\",\"end\"]\n",
    "oldPos[\"newStart\"]  = newPos.loc[:,\"start\"]\n",
    "oldPos[\"newEnd\"]  = newPos.loc[:,\"end\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>newStart</th>\n",
       "      <th>newEnd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16717</td>\n",
       "      <td>16744</td>\n",
       "      <td>16717</td>\n",
       "      <td>16744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>26454</td>\n",
       "      <td>26465</td>\n",
       "      <td>26454</td>\n",
       "      <td>26465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28589</td>\n",
       "      <td>28603</td>\n",
       "      <td>28589</td>\n",
       "      <td>28603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30863</td>\n",
       "      <td>30959</td>\n",
       "      <td>30863</td>\n",
       "      <td>30959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31720</td>\n",
       "      <td>31733</td>\n",
       "      <td>31720</td>\n",
       "      <td>31733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586489</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59030784</td>\n",
       "      <td>59030801</td>\n",
       "      <td>56884637</td>\n",
       "      <td>56884654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586490</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59032078</td>\n",
       "      <td>59032094</td>\n",
       "      <td>56885931</td>\n",
       "      <td>56885947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586491</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59032243</td>\n",
       "      <td>59032253</td>\n",
       "      <td>56886096</td>\n",
       "      <td>56886106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586492</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59033113</td>\n",
       "      <td>59033127</td>\n",
       "      <td>56886966</td>\n",
       "      <td>56886980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586493</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59033259</td>\n",
       "      <td>59033288</td>\n",
       "      <td>56887112</td>\n",
       "      <td>56887141</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1586494 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHR     start       end  newStart    newEnd\n",
       "0        chr1     16717     16744     16717     16744\n",
       "1        chr1     26454     26465     26454     26465\n",
       "2        chr1     28589     28603     28589     28603\n",
       "3        chr1     30863     30959     30863     30959\n",
       "4        chr1     31720     31733     31720     31733\n",
       "...       ...       ...       ...       ...       ...\n",
       "1586489  chrY  59030784  59030801  56884637  56884654\n",
       "1586490  chrY  59032078  59032094  56885931  56885947\n",
       "1586491  chrY  59032243  59032253  56886096  56886106\n",
       "1586492  chrY  59033113  59033127  56886966  56886980\n",
       "1586493  chrY  59033259  59033288  56887112  56887141\n",
       "\n",
       "[1586494 rows x 5 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oldPos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(range(1,23))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make sure the file is bgzipped and indexed\n",
    "(/storage/ydong/data/Filter_Merged_STRs_All_Samples_New.vcf.gz)\n",
    "\n",
    "bgzip file.vcf\n",
    "\n",
    "tabix -p vcf file.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GetGT(gt):\n",
    "    #print gt\n",
    "    if gt is None: return None\n",
    "    if gt == \"\": return None\n",
    "    if str(gt) == \".\" or str(gt) == \"./.\": return None #Missing genotype control\n",
    "    if \"|\" in gt:\n",
    "        gt = gt.split(\"|\")\n",
    "        return [int(gt[0]),int(gt[1])]\n",
    "    elif \"/\" in gt: \n",
    "        gt = gt.split(\"/\")\n",
    "        return [int(gt[0]),int(gt[1])]\n",
    "    else: \n",
    "        return [int(float(gt)), int(float(gt))] # haploid\n",
    "    \n",
    "def checkgt(record,s):\n",
    "    try:\n",
    "        return(record.genotype(s)[\"GB\"])\n",
    "    except:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPos(chrom,start,end):\n",
    "    curr = oldPos[oldPos[\"start\"] == start]\n",
    "    if len(curr) == 1:\n",
    "        return int(curr[\"newStart\"])\n",
    "    else:\n",
    "        curr = curr[curr[\"CHR\"] == chrom]\n",
    "        if len(curr) == 1:\n",
    "            return int(curr[\"newStart\"])\n",
    "        else:\n",
    "            curr = curr[curr[\"end\"]== end]\n",
    "            return int(curr[\"newStart\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHR 2 done\n",
      "CHR 3 done\n",
      "CHR 4 done\n",
      "CHR 5 done\n",
      "CHR 6 done\n",
      "CHR 7 done\n",
      "CHR 8 done\n",
      "CHR 9 done\n",
      "CHR 10 done\n",
      "CHR 11 done\n",
      "CHR 12 done\n",
      "CHR 13 done\n",
      "CHR 14 done\n",
      "CHR 15 done\n",
      "CHR 16 done\n",
      "CHR 17 done\n",
      "CHR 18 done\n",
      "CHR 19 done\n",
      "CHR 20 done\n",
      "CHR 21 done\n",
      "CHR 22 done\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "MINCOUNT = 1\n",
    "MINGENOTYPES = 5\n",
    "MINSAMPLES = 50\n",
    "\n",
    "chrRange = list(range(2,23))\n",
    "\n",
    "for CHR in chrRange:\n",
    "    CHR = str(CHR)\n",
    "\n",
    "    OUTFILE = \"/storage/ydong/data/STRs/\" + CHR + \".csv\"\n",
    "    \n",
    "    # fectch the chromosome we want to extract\n",
    "    vcf_reads =vcf_reader.fetch(CHR)\n",
    "    SAMPLES = vcf_reads.samples\n",
    "\n",
    "\n",
    "    # open out file and write header\n",
    "    out = open(OUTFILE, \"w\")\n",
    "    out.write(\"\\t\".join([\"chrom\",\"start\"] + SAMPLES)+\"\\n\")\n",
    "    \n",
    "    \n",
    "    counter = 0\n",
    "    \n",
    "    for record in vcf_reads:\n",
    "        chrom = record.CHROM\n",
    "        if \"chr\" not in chrom: chrom = \"chr%s\"%chrom\n",
    "        start = record.POS\n",
    "        end = record.INFO[\"END\"]\n",
    "        # update position\n",
    "        \n",
    "        \n",
    "        pos = getPos(chrom,start,end)\n",
    "\n",
    "        genotypes = [GetGT(checkgt(record,s)) for s in SAMPLES]\n",
    "\n",
    "\n",
    "        counters = {\n",
    "            \"numloci\": 0,\n",
    "            \"minmaf\": 0,\n",
    "            \"minsamples\": 0,\n",
    "            \"mingenotypes\":0,\n",
    "            \"novar\": 0,\n",
    "            \"keepers\": 0\n",
    "            }\n",
    "\n",
    "\n",
    "        # This section is checking for minimum num of genotypes and counts/genotype\n",
    "        gt_sum = [sum(gt) if gt is not None else None for gt in genotypes]\n",
    "        unique_vals = pd.Series(gt_sum).value_counts(dropna=True)\n",
    "        good_gts = unique_vals[unique_vals >= MINCOUNT].index\n",
    "        if len(good_gts) < MINGENOTYPES:\n",
    "            counters[\"mingenotypes\"] = counters[\"mingenotypes\"] + 1\n",
    "            continue\n",
    "        genotypes = [genotypes[i] if gt_sum[i] in good_gts else None for i in range(len(gt_sum))]\n",
    "        if len([item for item in genotypes if item is not None]) < MINSAMPLES:\n",
    "            counters[\"minsamples\"] = counters[\"minsamples\"] + 1\n",
    "            continue\n",
    "\n",
    "        for i in range(len(genotypes)):\n",
    "            curr = genotypes[i]\n",
    "            if curr == None:\n",
    "                genotypes[i] = \"NA,NA\"\n",
    "            else:\n",
    "                genotypes[i] = str(curr[0]) + \",\" + str(curr[1])    \n",
    "        out.write(\"\\t\".join(map(str,[chrom,pos]+genotypes))+\"\\n\")\n",
    "        counter = counter + 1\n",
    "        \n",
    "\n",
    "    out.close() \n",
    "    print(\"CHR \" + CHR + \" done\")\n",
    "    \n",
    "end = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
