{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We only perform our analysis on chr1 ~ chr 22, as sex chromosome expression varies between male and female, would lead to complex pipelines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['Unnamed: 0'] not found in axis\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-4225ad16aecb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    143\u001b[0m         \u001b[0mjunction_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjunction_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mjunction_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Chr\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCHROM\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0mexon_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexon_df_t\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannot_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'probe.id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 145\u001b[0;31m         \u001b[0mexon_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexon_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Unnamed: 0\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maxis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    146\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    147\u001b[0m         \u001b[0;31m#removing duplicated samples by taking highest coverage (in exon expression) sample\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4115\u001b[0m             \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4116\u001b[0m             \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minplace\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4117\u001b[0;31m             \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4118\u001b[0m         )\n\u001b[1;32m   4119\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   3912\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32min\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3913\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3914\u001b[0;31m                 \u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_drop_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3915\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3916\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_drop_axis\u001b[0;34m(self, labels, axis, level, errors)\u001b[0m\n\u001b[1;32m   3944\u001b[0m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3945\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3946\u001b[0;31m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3947\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0maxis_name\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mnew_axis\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3948\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib64/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, errors)\u001b[0m\n\u001b[1;32m   5338\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5339\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0merrors\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5340\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"{} not found in axis\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5341\u001b[0m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5342\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['Unnamed: 0'] not found in axis\""
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Compute Percent Spliced In (PSI)\n",
    "# PSI_df: the output matrix of psi\n",
    "# exon_gene_df: exon read count matrix\n",
    "# junction_gene_df: Exon-exon junction read count matrix\n",
    "# annot_gene_df: annotation file for exons\n",
    "def compute_gene_PSI(PSI_df,exon_gene_df,junction_gene_df,annot_gene_df,read_len):\n",
    "     \n",
    "    exons = list(exon_gene_df.index)\n",
    "    min_reads = 10\n",
    "    \n",
    "    # loop through each exon\n",
    "    for exon in exons:\n",
    "\n",
    "        exon_chrom = annot_gene_df.loc[exon]['probe.chr']\n",
    "        exon_start = annot_gene_df.loc[exon]['probe.start']\n",
    "        exon_stop = annot_gene_df.loc[exon]['probe.stop']\n",
    "        exon_len = math.fabs(exon_start - exon_stop) + 1\n",
    "        exon_strand = annot_gene_df.loc[exon]['probe.strand']\n",
    "        if exon_strand == '-':\n",
    "            orig_start = exon_start\n",
    "            exon_start = exon_stop\n",
    "            exon_stop = orig_start\n",
    "        \n",
    "        \n",
    "        #Get the A,B,C reads for PSI calculation\n",
    "        B_intervals = []\n",
    "        C_intervals = []\n",
    "        for interval in junction_gene_df.index:\n",
    "            chrom,start,stop = interval.split('_')\n",
    "            chrom = chrom.split(\"chr\")[1]\n",
    "            if str(chrom) != str(exon_chrom):\n",
    "                print(chrom)\n",
    "                print(exon_chrom)\n",
    "                print('Chromosomes aren\\'t matching, something is wrong!!')\n",
    "\n",
    "            if int(start) < exon_start and int(stop) > exon_stop:#C\n",
    "                C_intervals.append(interval)\n",
    "            if int(start) < exon_start and (int(stop) >= exon_start and int(stop) <= exon_stop):#B1\n",
    "                B_intervals.append(interval)\n",
    "            if (int(start) >= exon_start and int(start) <= exon_stop) and int(stop) > exon_stop:#B2\n",
    "                B_intervals.append(interval)\n",
    "\n",
    "        A_reads = exon_gene_df.loc[exon]\n",
    "        B_reads = junction_gene_df.loc[B_intervals]\n",
    "        B_reads = B_reads.iloc[:,2:].sum()\n",
    "\n",
    "        C_reads = junction_gene_df.loc[C_intervals]\n",
    "        C_reads = C_reads.iloc[:,2:].sum()\n",
    "\n",
    "        total_reads = A_reads + B_reads + C_reads\n",
    "        if len(total_reads[total_reads < min_reads])/len(total_reads) > 0.2:\n",
    "            continue\n",
    "        else:\n",
    "            A_B_norm = (A_reads + B_reads)/(read_len + exon_len-1)\n",
    "            C_norm = C_reads/(read_len-1)\n",
    "\n",
    "            mask = (A_B_norm == 0) & (C_norm == 0)\n",
    "            PSI_norm = 100*((A_B_norm)/(A_B_norm + C_norm)) if (A_B_norm.sum() + C_norm.sum()) > 0.00 else 0\n",
    "            PSI_norm = PSI_norm.where(~mask,0)\n",
    "            PSI_df[exon] = PSI_norm\n",
    "\n",
    "    return PSI_df\n",
    "\n",
    "\n",
    "# find out the maximum coverage sample for each individual\n",
    "def max_coverage_samples(exon_df):\n",
    "    'Given exon expression dataframe'\n",
    "    'Take first two elements of sample name'\n",
    "    'If there is duplicate sample'\n",
    "    'Return list of max coverage non-dup samples'\n",
    "    \n",
    "    med_exon_df = pd.DataFrame(exon_df.median(axis=0))\n",
    "    med_exon_df['samples'] = [\"-\".join(i.split('-')[:2]) for i in med_exon_df.index]\n",
    "    med_exon_df['original_samples'] = med_exon_df.index\n",
    "    highest_cov_samples = list(med_exon_df.groupby('samples').idxmax()[0]) \n",
    "    \n",
    "    return highest_cov_samples\n",
    "\n",
    "def filter_exons(PSI_df):\n",
    "    'Given PSI dataframe'\n",
    "    'Filter out exons with little variation:'\n",
    "    'PSI = 100% or PSI = 0% in >90% patients'\n",
    "    'Return filtered PSI'\n",
    "    \n",
    "    PSI_filtered_df = PSI_df.dropna(axis=1)\n",
    "    exons = PSI_filtered_df.columns\n",
    "    exons_to_remove = []\n",
    "    num_samples = len(PSI_df.index)\n",
    "    for exon in exons:\n",
    "        exon_PSI = PSI_filtered_df[exon]\n",
    "        num_zero = len(exon_PSI[exon_PSI == 0])\n",
    "        num_hundred = len(exon_PSI[exon_PSI == 100])\n",
    "        if num_zero/num_samples > 0.9 or num_hundred/num_samples > 0.9:\n",
    "            exons_to_remove.append(exon)\n",
    "    PSI_filtered_df = PSI_filtered_df.drop(exons_to_remove,axis=1).dropna()\n",
    "    new_index = [\"-\".join(i.split('-')[:2]) for i in PSI_filtered_df.index]\n",
    "    PSI_filtered_df.index = new_index\n",
    "\n",
    "    return PSI_filtered_df\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "#     parser = argparse.ArgumentParser(description=\"Compute Percent Spliced In (PSI) for any exon\")\n",
    "#     parser.add_argument(\"--exon_expr\", help=\"File path to exon expression dataset\", type=str, required=True)\n",
    "#     parser.add_argument(\"--junction_expr\", help=\"File path to junction expression dataset\", type=str, required=True)\n",
    "#     parser.add_argument(\"--exprannot\", help=\"Exon expression annotation file\", type=str, required=True)\n",
    "#     parser.add_argument(\"--chrom\", help=\"Restrict analysis to this chromosome\", type=str, required=False)\n",
    "#     parser.add_argument(\"--min_samples\", help=\"Require data for this many samples\", type=int, default=0)\n",
    "#     parser.add_argument(\"--read_len\", help=\"Paired end length, default is GTEx read length: 75bp \", type=int, default=75)\n",
    "#     parser.add_argument(\"--out\", help=\"Write data files to this file\", type=str, required=True)\n",
    "\n",
    "\n",
    "#     args = parser.parse_args()\n",
    "    EXONFILE = '/storage/ydong/data/gtexRNA/lung/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads_Parsed.csv'#args.exon_expr\n",
    "    JUNCTIONFILE = '/storage/ydong/data/gtexRNA/lung/GTEx_Analysis_2017-06-05_v8_STARv2.5.3a_junctions_Parsed.csv'#args.junction_expr\n",
    "    ANNOTFILE = '/storage/ydong/data/gtexRNA/gencode_gene_annotations_GRCh38.csv'#args.exprannot\n",
    "    \n",
    "    \n",
    "    # read in files\n",
    "    exon_df_t = pd.read_csv(EXONFILE,index_col = \"Name\")\n",
    "    annot_df_t = pd.read_csv(ANNOTFILE,index_col = 0,dtype = {\"probe.chr\":str})\n",
    "    junction_df_t = pd.read_csv(JUNCTIONFILE,index_col = \"Name\",dtype = {\"Chr\":str})\n",
    "    \n",
    "    chrList = list(range(2,23))\n",
    "    \n",
    "    for CHROM in chrList:\n",
    "#         CHROM = 1 #args.chrom\n",
    "        MINSAMPLES = 0 #args.min_samples\n",
    "        READLEN = 2*75 #args.read_len\n",
    "        OUTFILE = '/storage/ydong/data/chrom/lung/chr' + str(CHROM) + '_PSI.csv'#args.out\n",
    "\n",
    "        # format CHROM\n",
    "        CHROM = str(CHROM)\n",
    "        annot_df = annot_df_t[annot_df_t['probe.chr'] ==  str(CHROM)]\n",
    "        junction_df =junction_df_t[junction_df_t['Description'].isin(annot_df['gene.id'])]\n",
    "        junction_df = junction_df[junction_df[\"Chr\"] == str(CHROM)]\n",
    "        exon_df = exon_df_t.loc[list(annot_df['probe.id'])]\n",
    "\n",
    "        #removing duplicated samples by taking highest coverage (in exon expression) sample\n",
    "\n",
    "        highest_cov_samples = max_coverage_samples(exon_df) \n",
    "        exon_df = exon_df[highest_cov_samples]\n",
    "        highest_cov_samples[:0] = ['Description','Chr']\n",
    "        junction_df = junction_df[highest_cov_samples]\n",
    "        PSI_df = pd.DataFrame(index=exon_df.columns, columns=exon_df.index)\n",
    "        genes = set(annot_df['gene.id'])\n",
    "        print('Chromosome: {}'.format(CHROM))\n",
    "        print('Exons total: {}'.format(len(exon_df.index)))\n",
    "\n",
    "        for gene in genes:\n",
    "            exon_gene_df = exon_df.loc[exon_df.index.str.startswith(gene)]\n",
    "            junction_gene_df = junction_df[junction_df['Description'] == gene]\n",
    "            annot_gene_df = annot_df[annot_df['gene.id'] == gene]\n",
    "            annot_gene_df.index = annot_gene_df['probe.id']\n",
    "            PSI_df = compute_gene_PSI(PSI_df,\n",
    "                                      exon_gene_df,\n",
    "                                      junction_gene_df,\n",
    "                                      annot_gene_df,\n",
    "                                      READLEN)\n",
    "        PSI_filtered_df = filter_exons(PSI_df.dropna(axis=1))\n",
    "        print('Exons with PSI: {}\\n'.format(len(PSI_filtered_df.columns)))\n",
    "        PSI_filtered_df.to_csv(OUTFILE)\n",
    "        print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(range(2,23))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exons with PSI: 3686\n",
      "\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "for gene in genes:\n",
    "    exon_gene_df = exon_df.loc[exon_df.index.str.startswith(gene)]\n",
    "    junction_gene_df = junction_df[junction_df['Description'] == gene]\n",
    "    annot_gene_df = annot_df[annot_df['gene.id'] == gene]\n",
    "    annot_gene_df.index = annot_gene_df['probe.id']\n",
    "    PSI_df = compute_gene_PSI(PSI_df,\n",
    "                              exon_gene_df,\n",
    "                              junction_gene_df,\n",
    "                              annot_gene_df,\n",
    "                              READLEN)\n",
    "PSI_filtered_df = filter_exons(PSI_df.dropna(axis=1))\n",
    "print('Exons with PSI: {}\\n'.format(len(PSI_filtered_df.columns)))\n",
    "PSI_filtered_df.to_csv(OUTFILE)\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
